app/services/cypher_generator.py

```diff
@@ -64,65 +64,85 @@ def run_query(self, query_code):
-        all_valid = True
         # node_dict = {node['node_id']: node for node in nodes}
 
-        if all_valid:
-            match_clauses = []
-            return_clauses = []
-
-            # Track nodes that are included in relationships
-            used_nodes = set()
-            if not predicates:
-                # Case when there are no predicates
-                for node in nodes:
-                    var_name = f"n_{node['node_id']}"
-                    match_clauses.append(self.match_node(node, var_name))
-                    return_clauses.append(var_name)
-            else:
-                for i, predicate in enumerate(predicates):
-                    predicate_type = predicate['type'].replace(" ", "_")
-                    source_node = node_map[predicate['source']]
-                    target_node = node_map[predicate['target']]
-
-                    if i == 0:
-                        source_var = 's0'
-                        source_match = self.match_node(source_node, source_var)
-                        match_clauses.append(source_match)
-                    else:
-                        source_var = f"t{i-1}"
-
-                    target_var = f"t{i}"
-                    target_match = self.match_node(target_node, target_var)
-
-                    match_clauses.append(f"({source_var})-[r{i}:{predicate_type}]->{target_match}")
-                    return_clauses.append(f"r{i}")
-
-                    used_nodes.add(predicate['source'])
-                    used_nodes.add(predicate['target'])
-                
-                for node_id, node in node_map.items():
-                    if node_id not in used_nodes:
-                        var_name = f"n_{node_id}"
-                        match_clauses.append(self.match_node(node, var_name))
-                        return_clauses.append(var_name)
-
-                return_clauses.extend([f"s0"] + [f"t{i}" for i in range(len(predicates))])
-
-            match_clause = "MATCH " + ", ".join(match_clauses)
-            return_clause = "RETURN " + ", ".join(return_clauses)
-            cypher_query = f"{match_clause} {return_clause}"
+        match_preds = []
+        return_preds = []
+        match_no_preds = []
+        return_no_preds = []
+    
+        # Track nodes that are included in relationships
+        used_nodes = set()
+        if not predicates:
+            # Case when there are no predicates
+            for node in nodes:
+                var_name = f"n_{node['node_id']}"
+                match_no_preds.append(self.match_node(node, var_name))
+                return_no_preds.append(var_name)
+            cypher_query = self.construct_clause(match_no_preds, return_no_preds)
             cypher_queries.append(cypher_query)
         else:
-            logging.debug("Processing stopped due to invalid request.")
+            for i, predicate in enumerate(predicates):
+                predicate_type = predicate['type'].replace(" ", "_")
```

Comment:
To avoid potential case-sensitivity issues, it's recommended to apply `.lower()` to the predicate.

Reply:
done